Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs

```diff
@@ -194,24 +249,38 @@ void FixedUpdate()
             );
         }
 
-        // c. Encourage head height. //Should normalize to ~1
-        var headHeightOverFeetReward =
-            ((head.position.y - footL.position.y) + (head.position.y - footR.position.y) / 10);
-        if (float.IsNaN(headHeightOverFeetReward))
+        AddReward(matchSpeedReward * lookAtTargetReward);
+    }
+
+    //Returns the average velocity of all of the body parts
+    //Using the velocity of the hips only has shown to result in more erratic movement from the limbs, so...
+    //...using the average helps prevent this erratic movment
+    Vector3 GetAvgVelocity()
+    {
+        Vector3 velSum = Vector3.zero;
+        Vector3 avgVel = Vector3.zero;
+
+        //ALL RBS
+        int numOfRB = 0;
+        foreach (var item in m_JdController.bodyPartsList)
         {
-            throw new ArgumentException(
-                "NaN in headHeightOverFeetReward.\n" +
-                $" head.position: {head.position}\n"+
-                $" footL.position: {footL.position}\n"+
-                $" footR.position: {footR.position}"
-            );
+            numOfRB++;
+            velSum += item.rb.velocity;
         }
 
-        AddReward(
-            + 0.02f * moveTowardsTargetReward
-            + 0.02f * lookAtTargetReward
-            + 0.005f * headHeightOverFeetReward
-        );
+        avgVel = velSum / numOfRB;
+        return avgVel;
+    }
+
+    //normalized value of the difference in avg speed vs goal walking speed.
+    public float GetMatchingVelocityInverseLerp(Vector3 velocityGoal, Vector3 actualVelocity)
+    {
+        //distance between our actual velocity and goal velocity
+        var velDeltaMagnitude = Mathf.Clamp(Vector3.Distance(actualVelocity, velocityGoal), 0, walkingSpeed);
+
+        //return the value on a declining sigmoid shaped curve that decays from 1 to 0
+        //This reward will approach 1 if it matches perfectly and approach zero as it deviates
+        return Mathf.Pow(1 - Mathf.Pow(velDeltaMagnitude / walkingSpeed, 2), 2);
```

Comment:
Why do you need to square `velDeltaMagnitude / walkingSpeed` ? The value is already between 0 and 1.
It feels weird to me to use powers of 4 here.

Reply:
The reason is so that the x intercept happens at our targetWalkingSpeed ( I changed that variable name per your suggestion). You can visualize this function here: https://www.desmos.com/calculator/r52xqkfbny

Reply:
The function (1- (x/s))^2 also looks like it intercepts the X axis where we want

Reply:
no changes made. yeah looks like that would also work. i added it to that desmos graph. (your suggestion in red)
<img width="1177" alt="Screen Shot 2020-08-13 at 2 34 36 PM" src="https://user-images.githubusercontent.com/19273127/90189663-66448c00-dd72-11ea-8c43-c73a14d00e7a.png">

Reply:
My suggestion in green : 
<img width="624" alt="Screen Shot 2020-08-13 at 2 55 48 PM" src="https://user-images.githubusercontent.com/28320361/90191169-3c409900-dd75-11ea-828f-a50d4528b8b9.png">

Reply:
(1- (x/s))^2 =/= (1-(x/s)^2)